index.rst

```diff
@@ -1,29 +1,988 @@
+       )(normalize_datetime)
+
+       class Config:
+           json_encoders = {datetime: lambda v: int(v.timestamp())}
+
+Note the syntax for validating multiple fields with the same validator, and the Pydantic configuration for converting ``datetime`` back to seconds since epoch when returning the model as JSON (in, for example, a response body).
+
+Internal models
+---------------
+
+For models that are only used internally (such as between services and storage objects), prefer dataclasses_ to Pydantic models.
+Dataclasses are much simpler and signal that none of the complex validation done by Pydantic is in play.
+
+.. _dataclasses: https://docs.python.org/3/library/dataclasses.html
+
+As with Pydantic models, use Enum classes for any field that's limited to a specific set of values.
+
+Consider marking dataclasses as frozen and creating a new instance of the dataclass whenever you need to modify one.
+This makes them easier to reason about and avoids subtle bugs when dataclasses are stored in caches or other long-lived data structures.
+
+Methods on models
+-----------------
+
+Models, whether Pydantic or internal dataclasses, are intended only for carrying data from one part of the application to another.
+They should never be used to implement business logic or interact with external storage or user input (apart from validation rules).
+They are data structures and data containers, not repositories of code.
+
+The one case where methods on models are appropriate is for data conversion.
+Use custom constructors (written as class methods) to create a data model object by parsing some other representation of that object, and add methods such as ``to_dict`` or ``as_cookie`` to format the contents of the data model into some other representation.
+
+These methods should only do format conversion and input validation, not higher-level verification or business logic such as authorization checks.
+
+.. _handlers:
+
+Handlers
+========
+
+The purpose of a FastAPI handler is to convert an incoming web request into internal models, dispatch it to the services layer, and then format the response (if any) as a correct HTTP response.
+Ideally, as much of this as possible should be done by FastAPI rather than hand-written code.
+The ideal handler is two lines of code: ask the factory to create the relevant service object, and then call the service object with the input model, returning its result as the output model.
+
+The bulk of the handler should therefore be in the FastAPI decorator and in the parameter list.
+FastAPI generates the API documentation from that annotation, so make full use of all of the parameters that flesh out the documentation.
+Specifically, every handler should have a ``summary``, many handlers should have a ``responses`` parameter specifying their error codes and descriptions, many handlers should have a ``status_code`` parameter, and larger applications with a lot of handlers should use ``tags``.
```

Comment:
Just thinking we might need to write up a standard for tags so that Squareone can merge open API specs.

Reply:
Oh, yeah, that's a very good idea.